### PR TITLE
HSEARCH-2555 Ensure the test suite fail if any service isn't properly released

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-2555

This also fixes various issues with services not being released properly.

### CODE REVIEW

**Comment by yrodiere** [comment]
Jenkins, test this please

File: engine/src/main/java/org/hibernate/search/engine/service/impl/StandardServiceManager.java [hunk 1]
```diff
@@ -58,6 +62,7 @@ public StandardServiceManager(SearchConfiguration searchConfiguration,
 		this.providedServices = createProvidedServices( searchConfiguration );
 		this.defaultServices = defaultServices;
 		this.classloaderService = searchConfiguration.getClassLoaderService();
+		this.failOnUnreleasedService = Boolean.parseBoolean( System.getProperty( "org.hibernate.search.fail_on_unreleased_service" ) );
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/engine/service/impl/StandardServiceManager.java]: there's a short version for that:

    Boolean.getBoolean( "org.hibernate.search.fail_on_unreleased_service" );

File: engine/src/main/java/org/hibernate/search/engine/service/impl/StandardServiceManager.java [hunk 2]
```diff
@@ -108,10 +113,24 @@ public StandardServiceManager(SearchConfiguration searchConfiguration,
 
 	@Override
 	public void releaseAllServices() {
-		for ( ServiceWrapper wrapper : cachedServices.values() ) {
-			wrapper.ensureStopped();
+		List<String> unreleasedServicesToReport = failOnUnreleasedService ? new ArrayList<String>() : null;
+
+		for ( ServiceWrapper<?> wrapper : cachedServices.values() ) {
+			boolean stoppedProperly = wrapper.ensureStopped();
```
Comment (Sanne) [inline engine/src/main/java/org/hibernate/search/engine/service/impl/StandardServiceManager.java]: the wrapper needs to be synchronized here, like the original method which you're inlining now.
Failing to do so might expose you to a service instance to increment its usage counter during this stop operation for a legitimate usage, and yet throw the exception (or log the warning).

It's extremely unlikely as we'd be shutting down anyway.. but it's in theory possible if e.g. a service is needed to perform a clean teardown.
I'll fix that.

**Comment by Sanne** [comment]
merged, thanks! Just added two trivial commits

**Pull request closed without merging**